Add experiences timeline widget

kimlaberinto opened the pull request:
As discussed in #35 this is a PR to add a widget for a list of experiences containing title, organization, dates, location. Here are some example screenshots of the widget right now.

Widget background is `primary` in this case.
![image](https://user-images.githubusercontent.com/7938723/105941972-2cb3c480-6024-11eb-9043-06ab5423f53a.png)

Widget background is `secondary` in this case.
![image](https://user-images.githubusercontent.com/7938723/105942759-7c46c000-6025-11eb-91d2-b946008c482e.png)

**Features I could add:**
- Adding a draft=true/false for each experience, allowing the hugo theme user to write all their experiences in the `experiences.md` page and selecting whichever they want to show/hide.
- Adding a different style where it could just be listed in a plain format without the cards. 
- Adding an option to show Featured Images for each experiences
  - I have no idea how to do featured images in the front matter for each entries in front matter of `experiences.md`
  - I don't know how to use `utils/get-featured.html` for featured images typed in an entry of the front matter.

**Features that I omitted (for now?):**
- Maximum number of experiences to show, pagination, etc.
  - The user will have to manually manage which to experiences to show.
- FontAwesome icons for the Organization, Dates, and Location
  - I think it looks good and clean without the icons.

**Direct Questions:**
1. Right now the example `experiences` widget is at the bottom of the homepage, below the vintage widget. Any preference to the ordering of the widgets on the homepage?
2. Any feedback on style/look of the widget?
3. Any feedback on code?
4. Any tips for the features I might want to add?

wangchucheng commented on layouts/partials/widgets/experiences.html:
```diff
@@ -0,0 +1,20 @@
+{{ range $experiences }}
+    <div class="mb-6">
+        <div class="bg-secondary-bg rounded border hover:shadow-lg transition ease-in-out duration-200 px-6 pt-6 pb-4">
+            <div class="flex flex-col-reverse lg:flex-row justify-between">
+                <div class="w-full lg:w-2/3">
```
These two lines are not necessary I think. On the contrary, it will cause wrong styles, such as w-2/3.

kimlaberinto replied:
I see what you mean! I removed these two lines, and it fixed some width issues I was having. I will remove these as per your suggestion.

wangchucheng commented on layouts/partials/widgets/experiences.html:
```diff
@@ -0,0 +1,20 @@
+                    <div class="mb-4">
+                        <div class="font-bold text-xl">{{ .title }}</div>
+                        <div> {{ .organization | markdownify }}</div>
+                        <span class="mr-8"> {{ .dates | markdownify }}</span> <span> {{ .location | markdownify }} </span>
```
I think maybe we can narrow the gap between dates and location? Such as 4.

kimlaberinto replied:
> How about adding a link to organziation?
I think that would be a good idea. I can add an organization url to the front matter and hyperlink it appropriately.

> I think maybe we can narrow the gap between dates and location? Such as 4.
Yes, I think that would look good. I'll do 4 for the gap.

> And do you think it is a good idea to place organization, date, location in one line? Organization at left and the others at right.

Hmm I'm not sure. I would have to try it out.

wangchucheng reviewed (commented):
I think this style is great!

For your questions: 

1. Hugo will handle the `draft` automatically as far as I know. So you do not need to implement this.
2. A plain style is great if you want to add it!
3. For now, I think the experience should be kept concise and professional, so it is not necessary to add the feature image. Just like in a resume, pictures are generally not attached.
4. For now I think there is no need to paginate it and use icons. Just like what you did.
5. you can use the `weight` in front matter to change the order.

And please make sure to update contents in `resources/assets/css/css` if you use some styles new.

kimlaberinto commented:
@wangchucheng Here's the screenshot where the above card is with the fixed/adjusted style and using `mr-4` and the bottom is an altered version.

**Top:**
```
<div class="mb-4">
    <div class="font-bold text-xl">{{ .title }}</div>
    <div> {{ .organization | markdownify }}</div>
    <span class="mr-4"> {{ .dates | markdownify }}</span>
    <span> {{ .location | markdownify }} </span>
</div>
```

**Bottom:**
```
<div class="mb-4">
    <div class="font-bold text-xl">{{ .title }}</div>
    <a class="mr-4"> {{ .organization | markdownify }}</a>
    <span class="mr-4"> {{ .dates | markdownify }}</span>
    <span > {{ .location | markdownify }} </span>
</div>
```
![image](https://user-images.githubusercontent.com/7938723/105948969-88d11580-6031-11eb-9e94-05adb23eeb70.png)

If you mean to left justify the organization, and right justify the date and location that's also something we can try. Do you know how to do that with HTML/CSS? I'm not quite sure how to do so.

wangchucheng commented:
> If you mean to left justify the organization, and right justify the date and location that's also something we can try.

Yes, that's what I mean. You can use `flex` and `justify-between` to implement this. See [Justify Content](https://tailwindcss.com/docs/justify-content#space-between) for more details.

kimlaberinto commented:
@wangchucheng Here's how I implemented it and here's the screenshot.

The third code is:
```
<div class="mb-4">
    <div class="font-bold text-xl">{{ .title }}</div>
    <div class="flex justify-between">
        <a class=""> {{ .organization | markdownify }}</a>
        <div class=""> {{ .dates | markdownify }}</div>
        <div> {{ .location | markdownify }} </div>
    </div>
</div>
```
![image](https://user-images.githubusercontent.com/7938723/105950323-cafb5680-6033-11eb-8639-21d50b5ca86f.png)

I personally like the first or the second one.

wangchucheng commented:
What I mean is to right justify the date and location together. Maybe like this:

```
<div class="mb-4">
    <div class="font-bold text-xl">{{ .title }}</div>
    <div class="flex justify-between">
        <a class=""> {{ .organization | markdownify }}</a>
        <div>
            <div> {{ .dates | markdownify }}</div>
            <div> {{ .location | markdownify }} </div>
        </div>
    </div>
</div>
```

kimlaberinto commented:
Thank you for the code! That helped me out understand it better. Here's 3 more styles! 

Style 4 - (code you have above)

Style 5 - Location and date right justified.
```
<div class="mb-4">
    <div class="font-bold text-xl">{{ .title }}</div>
    <div class="flex justify-between">
        <a class=""> {{ .organization | markdownify }}</a>
        <div>
            <span class="mr-4"> {{ .dates | markdownify }}</span>
            <span> {{ .location | markdownify }} </span>
        </div>
    </div>
</div>
```

Style 6 - Organization and location left justified, dates right justified.
```
<div class="mb-4">
    <div class="font-bold text-xl">{{ .title }}</div>
    <div class="flex justify-between">
        <div>
            <a class="mr-4"> {{ .organization | markdownify }}</a>
            <span> {{ .location | markdownify }} </span>
        </div>
            <div class="mr-4"> {{ .dates | markdownify }}</div>
    </div>
</div>
```
![image](https://user-images.githubusercontent.com/7938723/105951252-59240c80-6035-11eb-9c06-f25912d9f950.png)


I really like Style 6 (the bottom one in the above screenshot)! I like the one line look! With the organization and location together since it makes sense for that information to be together.

If you agree with moving forward with Style 6, then I can probably add this to the code and we can probably do the merge tomorrow. Let me know what your preference is

By the way, do you have a preference to where the `experiences.html` widget should be on the homepage? I can change the weight to have it move up and down. Right now it's at the very bottom. I can leave it there if that's okay with you. Or I can move it just above `vintage`.

wangchucheng commented:
> I really like Style 6 (the bottom one in the above screenshot)! I like the one line look! With the organization and location together since it makes sense for that information to be together.

Yes, Style 6 is the best I think! And for the org link, how about the following config?

```
[organization]
  name = "Example Organization"
  url = "https://example.com/"
```

> I can move it just above `vintage`.

Yes, I agree with that.

And do you mind improve the documentation of this part? The docs is at [Hugo Eureka Docs](https://github.com/wangchucheng/hugo-eureka-docs). You can just wirte the English version and I will translate it to Chinese.

kimlaberinto commented:
Perfect!! Thank you for all the help. That format for the organization looks good! I'll put that in and let you know once its all good for review. Should be tomorrow.

And about the docs of course, no problem! I'll make sure to add docs for this widget.

Thanks again!

-------- Message d'origine --------
Le 27 janv. 2021 00 h 36, C. Wang < notifications@github.com > a écrit :
I really like Style 6 (the bottom one in the above screenshot)! I like the one line look! With the organization and location together since it makes sense for that information to be together.
Yes, Style 6 is the best I think! And for the org link, how about the following config?
[organization] name = "Example Organization" url = "https://example.com/"
I can move it just above vintage.
Yes, I agree with that.
And do you mind improve the documentation of this part? The docs is at Hugo Eureka Docs. You can just wirte the English version and I will translate it to Chinese.
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.￼
[ { "@context": "http://schema.org", "@type": "EmailMessage", "potentialAction": { "@type": "ViewAction", "target": "[https://github.com](<a href=)/wangchucheng/hugo-eureka/pull/37#issuecomment-768068480">https://github.com/wangchucheng/hugo-eureka/pull/37#issuecomment-768068480", "url": "[https://github.com](<a href=)/wangchucheng/hugo-eureka/pull/37#issuecomment-768068480">https://github.com/wangchucheng/hugo-eureka/pull/37#issuecomment-768068480", "name": "View Pull Request" }, "description": "View this Pull Request on GitHub", "publisher": { "@type": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

wangchucheng commented:
Thank you! How about if we change mr-4 to mr-2 for organization and/or add a separator like `·` between them?

And we can also remove mr-4 for dates.

```
            <div class="mb-4">
                <div class="font-bold text-xl">{{ .title }}</div>
                <div class="flex justify-between">
                    <div>
                        <a class="mr-4"> {{ .organization | markdownify }}</a>
                        <span> {{ .location | markdownify }} </span>
                    </div>
                        <div> {{ .dates | markdownify }}</div>
                </div>
            </div>
```

kimlaberinto commented:
Here are the changes, though there is an issue with smaller screen widths (e.g. mobile). Screenshots below.

Changes/Updates are the following: I added the seperator and spacing. I also added an option in the front matter to hide the specific experience entry if the user specifies (e.g. makes it easier to highlight which employment history is shown). I also added the organization name and url hyperlinking.

I don't know how to fix how the separator sometimes overflowing on smaller screens. There's also some issues with the location text being too close to the dates. I tried playing around with size breakpoints, `hidden` to deal with the separator issue but I couldn't get it to work. I was thinking maybe it could be `<div>` on mobile, and at the `sm:` size breakpoint it could become a `<span>` like normal.

**Normal Width**
![image](https://user-images.githubusercontent.com/7938723/106061633-43f1c100-60bb-11eb-9677-0e41b22d76ca.png)

**Smaller Width**
![image](https://user-images.githubusercontent.com/7938723/106062138-fd509680-60bb-11eb-8ec3-23d70315839c.png)

wangchucheng commented:
I have pushed a commit (c159ee9) to make it responsive.

And I think we can remove `hide_entry` Because people can just comment those experiences.

For the organization, do you think it is a good idea to use:

```
[organization]
  name = "Example Organization"
  url = "https://example.com/"
```

instead of:

```
    organization_name = "Example Organization 2"
    organization_url = ""
```

The first one is consistent with the form in about widget, but the drawback is that it may only be placed at the bottom to avoid parsing errors.

kimlaberinto commented:
Thank you for the commit to make it responsive. It looks good!

>And I think we can remove `hide_entry` Because people can just comment those experiences.

Okay, we can remove `hide_entry`. That's no problem! I removed it.

>Btw, I prefer .Params.something instead of $.Param "something". Although they are alomst the same.

Yeah that's fine. I'm new to Hugo so I wasn't sure how that worked. I changed it now.

>```
>[organization]
>name = "Example Organization"
>url = "https://example.com/"
>```
> the drawback is that it may only be placed at the bottom to avoid parsing errors

Personally, I would prefer that we stay with `organization_name` and `organization_url`. I see a few possible downsides:
- Having the organization info "far away" from the rest of entry information (writeup, location, etc.) is cumbersome/non-intuitive to the hugo theme user.
- Handling multiple organizations in the widget file might be an issue? I'm not sure how multiple organizations work with `[organization]', where the first entry has Organization A, and second entry has Organization B and so on.

While it's not a same style as the `about.md` file, I think the `organization_name` and `organization_url` is still easy/intuitive for users to understand.

Furthermore I don't want us to make the transition from `toml` to `yaml` since thats a change that seems like it should apply to all `.md` files. It feels too big of change and outside the scope of this PR. I think this widget is pretty much ready now for merging if that's okay with you. (The transition to `yaml` could be made in the future.)

wangchucheng commented:
> Furthermore I don't want us to make the transition from `toml` to `yaml` since thats a change that seems like it should apply to all `.md` files. It feels too big of change and outside the scope of this PR.

Yes! I totally agree that this transition is outside the scope of this PR. And what I intended to implement is like:

```
experiences:
  - title: Example Role 1
    dates: 2019 - Present
    location: Country
    organization:
        name: Example Organization 2
        url: 
    writeup: >
      Donec scelerisque egestas augue at tempor. Fusce faucibus magna in.
      - Suspendisse lacinia mauris a laoreet vehicula.
      - Aenean tincidunt enim vitae ante blandit tempor.
      - Nam tincidunt diam quis lorem rutrum ullamcorper.
```

I think this is intuitive to theme users and can be consistent with the about widget. Sadly, this syntax is only valid in yaml. 

As yaml is also a more widely used format, what I am thinking is that I will convert Eureka from toml to yaml format today or tomorrow. This format conversion will not affect existing users, because Hugo can handle both formats. After the conversion is complete, I will perform a simple code refactoring and merge this PR. What do you think of this method?

If you agree, I will merge this PR directly after completing these operations. If you have other ideas, we can continue the discussion here. In addition, I will be very grateful if you can provide the corresponding documentation for the widget when you have time. Thank you for your contribution!

kimlaberinto commented:
I really like that new format in yaml. I think that would be great for users!

>After the conversion is complete, I will perform a simple code refactoring and merge this PR. What do you think of this method?

Yeah that plan sounds great to me! Whenever you complete the transition, then feel free to refactor the code for `experiences.html`. Let me know if you run into any problems.

I'll definitely work on the documentation for the widget! I'm hoping this weekend I can put something together. Thanks again for all the help and for the theme!

wangchucheng commented:
I have merged this PR. After we finished the documentation, this feature will be available in the next release.
